### ISSUE TITLE
provide `files` API

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-metadata#15]
As far as i can see, we can't get `file` object directly by calling an API.

We instead  need call `listDir` or `getDir` to get `dir` object, which then contains `files` field.

While it mimics most of the filesystems, and might be OK for file system case, it is very inefficient to be used in object storage server like 0-stor or minio because to get object metadata, we need to do these:
- get list of (all?) files in a directory -> getting worse if we have many files in a dir
- iterate the `files` field of `dir` object -> it is a slice, not a `map`

provide `files` API to get metadata of a file directly, will save us from above unneeded inefficiencies

**Comment by lafont-e** [comment]
Another solution could be to change the file: [string] to file:[file] so instead of returning an array of file names it can return an array of file objects.

**Comment by lafont-e** [comment]
yes, I don't know why I was thinking that file was an array of strings

This does not help you in what you want? I understand that sometimes you are going to have a long path and as the api is currently, you need to traverse all to arrive, the question is to know how much times you are going to have the full path beforehand

**Comment by iwanbk** [comment]
> This does not help you in what you want?

It help, it only not efficient, i think.

> to know how much times you are going to have the full path beforehand

True

I also just realized that i can't find API to:
- add  file to dir
- remove file from dir

**Comment by iwanbk** [comment]
> I also just realized that i can't find API to:
>> add file to dir
>> remove file from dir

OK, i just found `updateDir` API. But it seems to be has same drawbacks as mentioned the issue description.

When we have a lot of files, then we will need to send (unnecessary) a lot of file objects on the wire

**Comment by zaibon** [comment]
All of this is by design. When you write something 90% of the time you know where to write it, so there is no need for multiple call to traverse the different directories.

Regarding sending too much info, I'm a bit worried about that too, maybe we could have some PUT methods to only update a small part of the Dir object.

@despiegk can you comment on this ?

**Comment by lafont-e** [comment]
I think the same, it's an excessively simple API, some methods to deal with files and dir would be needed, that or create an accompanying library that can supply the missing parts

**Comment by iwanbk** [comment]
> All of this is by design. When you write something 90% of the time you know where to write it

My concern is when we read the file. To read a file, we need to get metadata of all files in that directory.

**Comment by zaibon** [comment]
> My concern is when we read the file. To get metadata of a file, we need to get metadata of all files in that directory.

Well then this design if going to be fast since you get all info of the directory in one call

**Comment by lafont-e** [comment]
@iwanbk  maybe you will like to take a look at a proposal I have done,  (https://gist.github.com/lafont-e/b26b88c89d7734351a6db3fd9c648f21)
Would love your comments.
